source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -138,10 +149,42 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void ProxyFilter::onLoadDnsCacheComplete(const Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
-  ENVOY_STREAM_LOG(debug, "load DNS cache complete, continuing", *decoder_callbacks_);
+void ProxyFilter::addHostAddressToFilterState(
+    const Network::Address::InstanceConstSharedPtr& address) {
+
+  if (!config_->saveUpstreamAddress()) {
+    return;
+  }
+
+  if (!address) {
+    ENVOY_STREAM_LOG(debug, "Cannot add address to filter state: invalid address",
+                     *decoder_callbacks_);
+    return;
+  }
+
+  ENVOY_STREAM_LOG(trace, "Adding resolved host {} to filter state", *decoder_callbacks_,
+                   address->asString());
+
+  const Envoy::StreamInfo::FilterStateSharedPtr& filter_state =
+      decoder_callbacks_->streamInfo().filterState();
+
+  auto address_set = std::make_unique<StreamInfo::UpstreamAddressSet>();
+  address_set->addresses_.emplace(address);
+
+  filter_state->setData(StreamInfo::UpstreamAddressSet::key(), std::move(address_set),
+                        StreamInfo::FilterState::StateType::Mutable,
+                        StreamInfo::FilterState::LifeSpan::Request);
+}
+
+void ProxyFilter::onLoadDnsCacheComplete(
+    const Common::DynamicForwardProxy::DnsHostInfoSharedPtr& host_info) {
+  ENVOY_STREAM_LOG(debug, "load DNS cache complete, continuing after adding resolved host: {}",
+                   *decoder_callbacks_, host_info->resolvedHost());
   ASSERT(circuit_breaker_ != nullptr);
   circuit_breaker_.reset();
+
+  addHostAddressToFilterState(host_info->address());
```

Comment:
I think @ggreenway is saying to do a null check here, and don't call `addHostAddressToFilterState` unless it's not null. Then just assert it's non-null in `addHostAddressToFilterState`. I agree with this.

Reply:
> I think @ggreenway is saying to do a null check here, and don't call `addHostAddressToFilterState` unless it's not null. Then just assert it's non-null in `addHostAddressToFilterState`. I agree with this.

I could do that. A thought -  since the address being null is a problem only for adding it to FilterState, I think it might be better to localize that check in the `addHostAddressToFilterState` function.  It looks like address being null is not a problem in normal flow of things . @ggreenway WDYT?

Reply:
Sure, that would be fine with me. But whichever way you do it, I think it's worth a comment about the case(s) in which it can be nullptr.